esp/services/ws_workunits/ws_workunitsService.cpp

```diff
@@ -5014,3 +5014,442 @@ bool CWsWorkunitsEx::onWUGetArchiveFile(IEspContext &context, IEspWUGetArchiveFi
+}
+
+void CWsWorkunitsEx::doECLQuerySyntaxCheck(IEspContext &context, const char *target, const char *query, IEspWUECLQueryActionRequest &req, IArrayOf<IConstWUECLQueryActionResult> &results)
+{
+    Owned<IWorkUnitFactory> factory = getWorkUnitFactory(context.querySecManager(), context.queryUser());
+    NewWsWorkunit wu(factory, context);
+    wu->setAction(WUActionCheck);
+
+    MemoryBuffer mb;
+    fastLZCompressToBuffer(mb, 0, "");
+    StringBuffer text(mb.length(), mb.toByteArray());
+    wu.setQueryText(text.str());
+    wu.setQueryMain(query);
+
+    StringAttr wuid(wu->queryWuid());  // NB queryWuid() not valid after workunit.clear()
+    wu->commit();
+    wu.clear();
+
+    WsWuHelpers::submitWsWorkunit(context, wuid.str(), target, NULL, 0, true, false, false);
+    waitForWorkUnitToComplete(wuid.str(), req.getTimeToWait());
+
+    Owned<IConstWorkUnit> cw(factory->openWorkUnit(wuid.str()));
+    WUState st = cw->getState();
+    bool wuTimeout = (st != WUStateAborted) && (st != WUStateCompleted) && (st != WUStateFailed);
+    VStringBuffer result(" WUSyntaxCheckECL for %s:", query);
+    if (wuTimeout)
+        result.append(" timed out.");
+
+    readWUException(cw->getExceptions(), result);
+    addECLQueryActionResult(query, result.str(), "SyntaxCheck", true, results);
+    cw.clear();
+
+    if (wuTimeout)
+        abortWorkUnit(wuid.str(), context.querySecManager(), context.queryUser());
+    if (!factory->deleteWorkUnit(wuid.str()))
+    {
+        result.setf(" Workunit %s cannot be deleted now. You may delete it when its status changes.", wuid.str());
+        addECLQueryActionResult(query, result.str(), "SyntaxCheck", true, results);
+    }
+}
+
+bool CWsWorkunitsEx::deployECLQuery(IEspContext &context, const char *target, const char *name, StringBuffer &wuid, StringBuffer &result)
+{
+    Owned<CWUDeployWorkunitRequest> deployReq = new CWUDeployWorkunitRequest("WsWorkunits");
+    deployReq->setName(name);
+    deployReq->setQueryMainDefinition(name);
+    deployReq->setCluster(target);
+    deployReq->setFileName("");
+
+    MemoryBuffer mb;
+    fastLZCompressToBuffer(mb, 0, "");
+    deployReq->setObject(mb);
+    deployReq->setObjType("compressed_ecl_text");
+
+    Owned<CWUDeployWorkunitResponse> deployResponse = new CWUDeployWorkunitResponse("WsWorkunits");
+    onWUDeployWorkunit(context, *deployReq, *deployResponse);
+    return readDeployWUResponse(deployResponse, wuid, result);
+}
+
+void CWsWorkunitsEx::doECLQueryDeployment(IEspContext &context, const char *target,  const char *query, IArrayOf<IConstWUECLQueryActionResult> &results)
+{
+    StringBuffer wuid, finalResult;
+    deployECLQuery(context, target, query, wuid, finalResult);
+    addECLQueryActionResult(query, finalResult.str(), "Deploy", true, results);
+}
+
+void CWsWorkunitsEx::doECLQueryPublish(IEspContext &context, const char *target,  const char *query, IEspWUECLQueryActionRequest &req, IArrayOf<IConstWUECLQueryActionResult> &results)
```

Comment:
This is where calling it "query" will really get confusing.  Maybe:

void CWsWorkunitsEx::doPublishEclDefinition(IEspContext &context, const char *target,  const char *definition, IEspWUECLQueryActionRequest &req, IArrayOf<IConstWUECLQueryActionResult> &results);

void CWsWorkunitsEx::doDeployEclDefinition(IEspContext &context, const char *target,  const char *definition, IArrayOf<IConstWUECLQueryActionResult> &results);

Etc...

Reply:
For everything related...

 deployECLDefinition(context, target, definition, wuid, finalResult);

etc.